projects/components/src/filtering/filter-bar/filter-chip/filter-chip.service.ts

```diff
@@ -0,0 +1,130 @@
+import { Injectable } from '@angular/core';
+import { FilterBuilderLookupService } from '../../filter/builder/filter-builder-lookup.service';
+import { AbstractFilterBuilder } from '../../filter/builder/types/abstract-filter-builder';
+import { IncompleteFilter } from '../../filter/filter';
+import { FilterAttribute } from '../../filter/filter-attribute';
+import { FilterOperator } from '../../filter/filter-operators';
+import { FilterParserLookupService } from '../../filter/parser/filter-parser-lookup.service';
+import { SplitFilter } from '../../filter/parser/parsed-filter';
+import { AbstractFilterParser, splitFilterStringByOperator } from '../../filter/parser/types/abstract-filter-parser';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class FilterChipService {
+  public constructor(
+    private readonly filterBuilderLookupService: FilterBuilderLookupService,
+    private readonly filterParserLookupService: FilterParserLookupService
+  ) {}
+
+  public autocompleteFilters(attributes: FilterAttribute[], text: string = ''): IncompleteFilter[] {
+    return attributes
+      .filter(attribute => this.filterBuilderLookupService.isBuildableType(attribute.type))
+      .flatMap(attribute => this.autocompleteFilter(attribute, text))
+      .filter(incompleteFilter => this.filterMatchesUserText(text, incompleteFilter));
+  }
+
+  private filterMatchesUserText(text: string, incompleteFilter: IncompleteFilter): boolean {
+    const isStringMatch = incompleteFilter.userString.toLowerCase().includes(text.trim().toLowerCase());
+
+    if (incompleteFilter.operator === undefined) {
+      return isStringMatch;
```

Comment:
If `isStringMatch` is true and operator is defined, what does that mean? When would a string match ever not be good enough? Or is this just meant to shortcircuit so lookup doesn't fail on the undefined? If that's the case, I'd shortcircuit on a string match too.

Reply:
Yeah, this is a bit confusing. Back in simpler times, the `isStringMatch` could have been the only check. This changed with the new syntax. Going to add the following comment here:

```
    /*
     * In most cases, the isStringMatch should be the only check that is needed, however this check fails for
     * STRING_MAP types with an operator since the LHS includes a user entered key between the attribute name
     * and the operator. We fix this by sending the full string through our parsing logic to see if we can pull
     * the key value off the LHS and build a filter out of it. If so, its a valid match and we want to include
     * it as an autocomplete option.
     */
```

Let me know if that makes sense.

Reply:
It does - logic organization wise though, I'd still recommend returning immediately if isStringMatch is true - it just convey the intent much better.

Reply:
Fixed.